File: src/s3.ts
import { mkdir } from "node:fs/promises";
import { basename, dirname, join, relative, resolve } from "node:path";
import { S3Client } from "bun";
import type { S3Options } from "./config.ts";

export type S3Object = {
  key: string;
  size: number;
  lastModified?: Date;
};

type S3Credentials = ConstructorParameters<typeof S3Client>[0];

function endpointFor(options: S3Options): string {
  const endpoint = options.endpoint.replace(/\/+$/, "");
  if (endpoint.includes("://")) {
    return options.useHttps
      ? endpoint.replace(/^http:\/\//i, "https://")
      : endpoint.replace(/^https:\/\//i, "http://");
  }
  return `${options.useHttps ? "https" : "http"}://${endpoint}`;
}

export function s3ObjectKey(dumpsRoot: string, filePath: string): string {
  const rel = relative(resolve(dumpsRoot), resolve(filePath));
  if (!rel || rel.startsWith("..") || rel.includes(`..${"/"}`)) {
    throw new Error("Selected dump must be inside the configured dumps directory");
  }
  return rel.split("\\").join("/");
}

function safeObjectKey(key: string): string {
  const normalized = key.replaceAll("\\", "/").replace(/^\/+/, "");
  if (
    !normalized ||
    normalized.split("/").some((part) => part === ".." || part === ".")
  ) {
    throw new Error(`Unsafe S3 object key: ${key}`);
  }
  return normalized;
}

export function localPathForS3Object(dumpsRoot: string, key: string): string {
  const safe = safeObjectKey(key);
  const root = resolve(dumpsRoot);
  const path = resolve(root, safe);
  if (path !== root && !path.startsWith(`${root}/`)) {
    throw new Error("S3 object would download outside the dumps directory");
  }
  return path;
}

export function createS3Client(
  options: S3Options,
  secretAccessKey: string,
): S3Client {
  const credentials = {
    bucket: options.bucketName,
    accessKeyId: options.accessKey,
    secretAccessKey,
    endpoint: endpointFor(options),
    region: options.region,
    // Bun uses virtual-hosted addressing by default. S3-compatible services
    // that require path-style addressing can still accept this option.
    virtualHostedStyle: !options.forcePathStyle,
  } as S3Credentials;
  return new S3Client(credentials);
}

export async function listS3Objects(
  options: S3Options,
  secretAccessKey: string,
): Promise<S3Object[]> {
  const credentials = {
    bucket: options.bucketName,
    accessKeyId: options.accessKey,
    secretAccessKey,
    endpoint: endpointFor(options),
    region: options.region,
  } as S3Credentials;
  const result = await S3Client.list(null, credentials);
  return (result.contents ?? [])
    .map((item) => ({
      key: item.key,
      size: item.size ?? 0,
      lastModified: item.lastModified
        ? new Date(item.lastModified)
        : undefined,
    }))
    .filter((item) => item.key.endsWith(".dump") || item.key.endsWith(".dump.enc"))
    .sort((a, b) => b.key.localeCompare(a.key));
}

export async function verifyS3Bucket(
  options: S3Options,
  secretAccessKey: string,
): Promise<void> {
  try {
    await listS3Objects(options, secretAccessKey);
  } catch (err) {
    if (options.createBucketIfNotExists) {
      throw new Error(
        `Cannot access S3 bucket "${options.bucketName}". Bun's native S3 API does not create buckets; create it first or disable createBucketIfNotExists. ${
          err instanceof Error ? err.message : String(err)
        }`,
      );
    }
    throw err;
  }
}

export async function uploadToS3(
  client: S3Client,
  dumpsRoot: string,
  filePath: string,
): Promise<string> {
  const key = s3ObjectKey(dumpsRoot, filePath);
  await client.file(key).write(Bun.file(filePath));
  return key;
}

export async function downloadFromS3(
  client: S3Client,
  dumpsRoot: string,
  key: string,
): Promise<string> {
  const localPath = localPathForS3Object(dumpsRoot, key);
  await mkdir(dirname(localPath), { recursive: true });
  await Bun.write(localPath, client.file(safeObjectKey(key)));
  return localPath;
}

export function formatS3Object(object: S3Object): string {
  const date = object.lastModified
    ? ` ${object.lastModified.toISOString()}`
    : "";
  return `${basename(object.key)} (${object.size} B)${date}`;
}

File: src/internal/s3/s3.go
package s3

import (
	"context"
	"fmt"
	"io"
	"net/url"
	"os"
	"path/filepath"
	"sort"
	"strings"
	"time"

	"github.com/aws/aws-sdk-go-v2/aws"
	"github.com/aws/aws-sdk-go-v2/credentials"
	"github.com/aws/aws-sdk-go-v2/service/s3"
	"github.com/md-redwan-hossain/dumpmgr/src/internal/config"
)

type Object struct {
	Key          string
	Size         int64
	LastModified *time.Time
}

func endpointFor(opts config.S3Options) (string, bool) {
	endpoint := strings.TrimRight(opts.Endpoint, "/")
	if strings.Contains(endpoint, "://") {
		if opts.UseHTTPS {
			endpoint = strings.Replace(endpoint, "http://", "https://", 1)
		} else {
			endpoint = strings.Replace(endpoint, "https://", "http://", 1)
		}
		return endpoint, opts.UseHTTPS
	}
	if opts.UseHTTPS {
		return "https://" + endpoint, true
	}
	return "http://" + endpoint, false
}

func NewClient(opts config.S3Options, secretAccessKey string) (*s3.Client, error) {
	endpoint, _ := endpointFor(opts)
	region := opts.Region
	if region == "" {
		region = "us-east-1"
	}
	cfg := aws.Config{
		Region:      region,
		Credentials: credentials.NewStaticCredentialsProvider(opts.AccessKey, secretAccessKey, ""),
	}
	client := s3.NewFromConfig(cfg, func(o *s3.Options) {
		o.BaseEndpoint = aws.String(endpoint)
		o.UsePathStyle = opts.ForcePathStyle
	})
	return client, nil
}

func ObjectKey(dumpsRoot, filePath string) (string, error) {
	root, err := filepath.Abs(dumpsRoot)
	if err != nil {
		return "", err
	}
	abs, err := filepath.Abs(filePath)
	if err != nil {
		return "", err
	}
	rel, err := filepath.Rel(root, abs)
	if err != nil || rel == "" || strings.HasPrefix(rel, "..") || strings.Contains(rel, ".."+string(filepath.Separator)) {
		return "", fmt.Errorf("selected dump must be inside the configured dumps directory")
	}
	return filepath.ToSlash(rel), nil
}

func safeObjectKey(key string) (string, error) {
	normalized := strings.TrimLeft(strings.ReplaceAll(key, "\\", "/"), "/")
	parts := strings.Split(normalized, "/")
	for _, part := range parts {
		if part == ".." || part == "." {
			return "", fmt.Errorf("unsafe S3 object key: %s", key)
		}
	}
	if normalized == "" {
		return "", fmt.Errorf("unsafe S3 object key: %s", key)
	}
	return normalized, nil
}

func LocalPathForObject(dumpsRoot, key string) (string, error) {
	safe, err := safeObjectKey(key)
	if err != nil {
		return "", err
	}
	root, err := filepath.Abs(dumpsRoot)
	if err != nil {
		return "", err
	}
	path := filepath.Join(root, filepath.FromSlash(safe))
	abs, err := filepath.Abs(path)
	if err != nil {
		return "", err
	}
	if abs != root && !strings.HasPrefix(abs, root+string(filepath.Separator)) {
		return "", fmt.Errorf("S3 object would download outside the dumps directory")
	}
	return abs, nil
}

func ListObjects(ctx context.Context, opts config.S3Options, secretAccessKey string) ([]Object, error) {
	client, err := NewClient(opts, secretAccessKey)
	if err != nil {
		return nil, err
	}
	paginator := s3.NewListObjectsV2Paginator(client, &s3.ListObjectsV2Input{
		Bucket: aws.String(opts.BucketName),
	})
	var objects []Object
	for paginator.HasMorePages() {
		page, err := paginator.NextPage(ctx)
		if err != nil {
			return nil, err
		}
		for _, item := range page.Contents {
			if item.Key == nil {
				continue
			}
			key := *item.Key
			if !strings.HasSuffix(key, ".dump") && !strings.HasSuffix(key, ".dump.enc") {
				continue
			}
			var lm *time.Time
			if item.LastModified != nil {
				t := *item.LastModified
				lm = &t
			}
			size := int64(0)
			if item.Size != nil {
				size = *item.Size
			}
			objects = append(objects, Object{
				Key:          key,
				Size:         size,
				LastModified: lm,
			})
		}
	}
	sort.Slice(objects, func(i, j int) bool {
		return objects[i].Key > objects[j].Key
	})
	return objects, nil
}

func VerifyBucket(ctx context.Context, opts config.S3Options, secretAccessKey string) error {
	_, err := ListObjects(ctx, opts, secretAccessKey)
	if err != nil && opts.CreateBucketIfNotExists {
		return fmt.Errorf(`cannot access S3 bucket %q. Bucket creation is not supported automatically; create it first or disable createBucketIfNotExists: %w`,
			opts.BucketName, err)
	}
	return err
}

func Upload(ctx context.Context, client *s3.Client, bucket, dumpsRoot, filePath string) (string, error) {
	key, err := ObjectKey(dumpsRoot, filePath)
	if err != nil {
		return "", err
	}
	f, err := os.Open(filePath)
	if err != nil {
		return "", err
	}
	defer f.Close()
	_, err = client.PutObject(ctx, &s3.PutObjectInput{
		Bucket: aws.String(bucket),
		Key:    aws.String(key),
		Body:   f,
	})
	if err != nil {
		return "", err
	}
	return key, nil
}

func Download(ctx context.Context, client *s3.Client, bucket, dumpsRoot, key string) (string, error) {
	safe, err := safeObjectKey(key)
	if err != nil {
		return "", err
	}
	localPath, err := LocalPathForObject(dumpsRoot, safe)
	if err != nil {
		return "", err
	}
	if err := os.MkdirAll(filepath.Dir(localPath), 0o755); err != nil {
		return "", err
	}
	out, err := client.GetObject(ctx, &s3.GetObjectInput{
		Bucket: aws.String(bucket),
		Key:    aws.String(safe),
	})
	if err != nil {
		return "", err
	}
	defer out.Body.Close()
	f, err := os.Create(localPath)
	if err != nil {
		return "", err
	}
	defer f.Close()
	if _, err := io.Copy(f, out.Body); err != nil {
		return "", err
	}
	return localPath, nil
}

func FormatObject(obj Object) string {
	date := ""
	if obj.LastModified != nil {
		date = " " + obj.LastModified.UTC().Format(time.RFC3339)
	}
	base := filepath.Base(obj.Key)
	return fmt.Sprintf("%s (%d B)%s", base, obj.Size, date)
}

func ParseEndpointHost(endpoint string) string {
	u, err := url.Parse(endpoint)
	if err == nil && u.Host != "" {
		return u.Host
	}
	return strings.TrimPrefix(strings.TrimPrefix(endpoint, "https://"), "http://")
}
